### ISSUE TITLE
quarkus-amazon-lambda fails with S3Event with native image

### ISSUE THREAD

**Issue by geniusit** [quarkusio/quarkus#7670]
**Describe the bug**
I'm trying to process a binary file in a AWS Lambda function. The lambda is triggered by a PUT / POST of the binary file in S3 (S3Event).

Everything works well on JVM Hot Spot. However it fails with native image with the following error : 

`
2020-03-07 01:33:50,132 ERROR [io.qua.ama.lam.run.AmazonLambdaRecorder] (Lambda Thread) Failed to run lambda: com.fasterxml.jackson.databind.JsonMappingException: Cannot set final field: com.amazonaws.services.s3.event.S3EventNotification.records. Enable by specifying "allowWrite" for this field in the reflection configuration.
`

So I created these two files : 

resources-config.json : 

```json{
  "resources": [
    {
      "pattern": ".*\\.json$"
    }
  ]
}
```

reflection-config.json :

```json
[
  {
    "name" : "com.amazonaws.services.s3.event.S3EventNotification",
    "allDeclaredConstructors" : true,
    "allPublicConstructors" : true,
    "allDeclaredMethods" : true,
    "allPublicMethods" : true,
    "allDeclaredFields" : true,
    "allPublicFields" : true,
    "fields" : [
      { "name" :  "records", "allowWrite":  true}
    ]
  }
]
```

Also add this line in my application.properties : 
```
quarkus.native.additional-build-args = -H:ResourceConfigurationFiles=resources-config.json, -H:ReflectionConfigurationFiles=reflection-config.json
```

The parameters are correctly passed to the the native image creation : 

```
native-image -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager --initialize-at-run-time=com.amazonaws.services.s3.event.S3EventNotification -H:ResourceConfigurationFiles=resources-config.json -H:ReflectionConfigurationFiles=reflection-config.json --initialize-at-build-time= -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy$BySpaceAndTime  ...
```
Here is the lambda code : 

```java
@Override
    public String handleRequest(S3Event event, Context context) {
        try {

            S3EventNotification.S3EventNotificationRecord record = event.getRecords().get(0);
            AmazonS3 s3client = AmazonS3ClientBuilder.defaultClient();
            S3Object s3object = s3client.getObject(record.getS3().getBucket().getName(), record.getS3().getObject().getUrlDecodedKey());
            S3ObjectInputStream inputStream = s3object.getObjectContent();

            service.process( inputStream );

            return "OK";
        } catch ( Throwable error ) {
            log.info("error " + error.getMessage());
            throw new RuntimeException( error );
        }
    }
```
But the error still occurs and I don't understand why.

**Expected behavior**
I would like to retrieve the record like I do with JVM mode

**Actual behavior**
I got the following exception : 
<code>
2020-03-07 14:27:02,752 ERROR [io.qua.ama.lam.run.AmazonLambdaRecorder] (Lambda Thread) Failed to run lambda: com.fasterxml.jackson.databind.JsonMappingException: Cannot set final field: com.amazonaws.services.s3.event.S3EventNotification.records. Enable by specifying "allowWrite" for this field in the reflection configuration.
 at [Source: (sun.net.www.protocol.http.HttpURLConnection$HttpInputStream); line: 1, column: 846] (through reference chain: com.amazonaws.services.lambda.runtime.events.S3Event["Records"])
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:611)
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty._throwAsIOE(SettableBeanProperty.java:599)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:152)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1719)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1228)
	at io.quarkus.amazon.lambda.runtime.AmazonLambdaRecorder$2.run(AmazonLambdaRecorder.java:150)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:497)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: java.lang.IllegalAccessException: Cannot set final field: com.amazonaws.services.s3.event.S3EventNotification.records. Enable by specifying "allowWrite" for this field in the reflection configuration.
	at java.lang.reflect.Field.set(Field.java:780)
	at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:150)
	... 8 more
</code>

**To Reproduce**
1. Create a quarkus project with the following dependencies : 
```xml
<dependencyManagement>
        <dependencies>
            <dependency>
                <groupId>io.quarkus</groupId>
                <artifactId>quarkus-bom</artifactId>
                <version>${quarkus.version}</version>
                <type>pom</type>
                <scope>import</scope>
            </dependency>
        </dependencies>
    </dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>com.amazonaws</groupId>
            <artifactId>aws-java-sdk-s3</artifactId>
            <version>1.11.738</version>
        </dependency>
        <dependency>
            <groupId>io.quarkus</groupId>
            <artifactId>quarkus-junit5</artifactId>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>28.2-jre</version>
        </dependency>
        <dependency>
              <groupId>io.quarkus</groupId>
              <artifactId>quarkus-amazon-lambda</artifactId>
          </dependency>
    </dependencies>
```

Notice that I added the : 
```xml
<dependency>
            <groupId>com.amazonaws</groupId>
            <artifactId>aws-java-sdk-s3</artifactId>
            <version>1.11.738</version>
        </dependency>
```
In order to work with the S3EventNotification class which is not part of quarkus. Actually, the error comes from this class. It seems that quarkus actually doestn't support S3EventNotification and only support JSON payload as input parameter. But my file is a binary so I need to retrieve it with the information that I get from the S3EventNotification object.
2. mvn clean package -Pnative
3. Create a lambda with a custom runtine env. Do not forget to add the DISABLE_SIGNAL_HANDLERS to true as environment variable.
4. Set a S3 bucket as source destination with PUT / POST object notification.
5. Upload a file to that bucket and see that the lambda is correctly triggered. However the exception is throwing saying that the records of the S3EventNotification objects cannot be set due to reflection problem.

**Configuration**
```properties
# Add your application.properties here, if applicable.
quarkus.lambda.handler=NatOtiJavaHandler
quarkus.index-dependency.s3.group-id = com.amazonaws
quarkus.index-dependency.s3.artifact-id = aws-java-sdk-s3
quarkus.index-dependency.joda-time.group-id = joda-time
quarkus.index-dependency.joda-time.artifact-id = joda-time
quarkus.native.additional-build-args =--initialize-at-run-time=com.amazonaws.services.s3.event.S3EventNotification -H:ResourceConfigurationFiles=resources-config.json, -H:ReflectionConfigurationFiles=reflection-config.json
quarkus.debug.reflection=true
```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)
![Screenshot from 2020-03-07 14-52-00](https://user-images.githubusercontent.com/3659037/76144696-4ef6c980-6083-11ea-8eac-cede4817413f.png)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
uname -a                                        
Linux localhost.localdomain 5.4.10-200.fc31.x86_64 #1 SMP Thu Jan 9 19:58:12 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

 - Output of `java -version`: 
openjdk version "1.8.0_242"
OpenJDK Runtime Environment Corretto-8.242.07.1 (build 1.8.0_242-b07)
OpenJDK 64-Bit Server VM Corretto-8.242.07.1 (build 25.242-b07, mixed mode)
 - GraalVM version (if different from Java): 19.3.1.r11-grl
 - Quarkus version or git rev: 1.2.1.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 
Apache Maven 3.6.0 (97c98ec64a1fdfee7767ce5ffb20918da4f719f3; 2018-10-24T20:41:47+02:00)

**Additional context**
(Add any other context about the problem here.)

**Comment by geniusit** [comment]
I'd precise that the constructor of the S3Event class is annotated by `@JsonCreator` as you can see here : 
```java
@JsonCreator
    public S3EventNotification(@JsonProperty("Records") List<S3EventNotification.S3EventNotificationRecord> records) {
        this.records = records;
    }
```

I suspect that the problem comes from there ...

**Comment by geniusit** [comment]
@patriot1burke Please could you take a look on it ? Thank you

**Comment by gsmet** [comment]
@geniusit could you try this PR: https://github.com/quarkusio/quarkus/pull/7694 ?

Get the code, then run `mvn clean install -DskipTests -DskipITs` then use `999-SNAPSHOT` as the version (and be sure to use `quarkus-bom` as the bom).

If it works for you, let's make a deal: you provide me a simple Jackson test case that is failing in native mode that I can include in the test suite.

**Comment by geniusit** [comment]
@gsmet I finally succeed to make my code work. 
First I successfully tested your PR but nothing has changed and still got the error.

Then I read this post : https://stackoverflow.com/questions/36730266/aws-lambda-s3event-deserialization
And that's the trick. Do not use S3Event but use S3EventNotofication in the signature.

Like this : 

```java 
@Override
    public String handleRequest(S3EventNotification event, Context context) {
        try {

            S3EventNotification.S3EventNotificationRecord record = event.getRecords().get(0);
            AmazonS3 s3client = AmazonS3ClientBuilder.defaultClient();
            S3Object s3object = s3client.getObject(record.getS3().getBucket().getName(), record.getS3().getObject().getUrlDecodedKey());
            S3ObjectInputStream inputStream = s3object.getObjectContent();

            service.process( inputStream );

            return "OK";
        } catch ( Throwable error ) {
            log.info("error " + error.getMessage());
            throw new RuntimeException( error );
        }
    }
```

So I removed my resources-config.json and reflection-config.json and the extra parameters that I sent to the native image creation command.

The only mandatory config is : 

quarkus.index-dependency.s3.group-id = com.amazonaws
quarkus.index-dependency.s3.artifact-id = aws-java-sdk-s3

quarkus.index-dependency.joda-time.group-id = joda-time
quarkus.index-dependency.joda-time.artifact-id = joda-time

If I don't set these parameters in my application.properties then I got the following warning at built time : 

WARNING] [io.quarkus.deployment.steps.ReflectiveHierarchyStep] Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
	- com.amazonaws.services.s3.event.S3EventNotification
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.");.

And the following error at runtime when invoking the AWS lambda : 

2020-03-10 10:16:29,395 ERROR [io.qua.ama.lam.run.AmazonLambdaRecorder] (Lambda Thread) Failed to run lambda: com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `com.amazonaws.services.s3.event.S3EventNotification$S3EventNotificationRecord` (no Creators, like default construct, exist): cannot deserialize from Object value (no delegate- or property-based Creator)

So I don't know if your PR has the be merged or not. The `@JsonCreator` is successfully managed by the Jandex index. There is no bug in quarkus related to my situation. 

This issue can be closed. 
Thank you again for support. Regards

**Comment by gsmet** [comment]
So do you need my PR in or not?

I'm surprised you need Joda to be indexed. What's the reason for that? The S3 jar uses it in serialized objects?

I think we might improve the experience by indexing these things automatically if they are needed in a lot of Lambda use cases.

**Comment by geniusit** [comment]
I dont need your PR because I solved my problem by changing the signature from S3Event to S3EventNotification.

For Joda indexation : 

First I need this maven dependency in order to work with the S3EventNotification object : 
```xml
<dependency>
    <groupId>com.amazonaws</groupId>
    <artifactId>aws-java-sdk-s3</artifactId>
    <version>1.11.738</version>
</dependency>
````

As soon as I add this dependency I must add these two lines as explained in my previous response : 

quarkus.index-dependency.s3.group-id = com.amazonaws
quarkus.index-dependency.s3.artifact-id = aws-java-sdk-s3

And as soon as I add these two lines I got the following warning at build time : 

WARNING] [io.quarkus.deployment.steps.ReflectiveHierarchyStep] Unable to properly register the hierarchy of the following classes for reflection as they are not in the Jandex index:
- org.joda.time.DateTime
Consider adding them to the index either by creating a Jandex index for your dependency via the Maven plugin, an empty META-INF/beans.xml or quarkus.index-dependency properties.");.

So this is why I must also add  : 

quarkus.index-dependency.joda-time.group-id = joda-time
quarkus.index-dependency.joda-time.artifact-id = joda-time

Notice that it's independant from the io.quarkus:quarkus-amazon-lambda module. It's S3 related.

May be that a new quarkus extensions can be requested : io.quarkus:quarkus-amazon-s3

Actually I think that the Camel S3 storage service component doesn't support S3EventNotification.
